### PR TITLE
Separate loop and plugin control

### LINKED ISSUE #52369: kube-controller-manager --cloud-provider conflates loops and plugins
/kind bug

**What happened**: The --cloud-provider flag controls both which control loops are run and which plugin is loaded. However the volume controller loop is still being run in the kube-controller-manager. The volume controller manager needs the appropriate plugin initialized if it is going to control things like GCE Persistent Disk or AWS Elastic Block Store.

**What you expected to happen**: We should have a second flag to separate the two behaviors we want to control. That way we run only the "external" control loops but still load the appropriate plugin. (gce_pd, aws_ebs, ...)

**How to reproduce it (as minimally and precisely as possible)**: Run "external" on a cloud provider with custom volumes and note they do not come up. 

**Anything else we need to know?**: Nope

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9
- Cloud provider or hardware configuration**: GCE
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by cheftako**
**What this PR does / why we need it**: Separate loop and plugin control in the kube-controller-manager.
Adding an "--external-plugin" flag to specify a plugin to load when
cloud-provider is set to "external". Flag has no effect currently
when the cloud-provider is not set to external. The expectation is
that the cloud provider and external plugin flags would go away once
all cloud providers are on stage 2 cloud-controller-manager solutions.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #52369 

**Special notes for your reviewer**:

**Release note**:
```release-note Added --external-plugin flag to kube-controller-manager to determine plugin when --cloud-provider is set to external.
```

### CODE REVIEW

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -131,6 +131,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet, allControllers []string, disabled
 	fs.BoolVar(&s.UseServiceAccountCredentials, "use-service-account-credentials", s.UseServiceAccountCredentials, "If true, use individual service account credentials for each controller.")
 	fs.StringVar(&s.CloudProvider, "cloud-provider", s.CloudProvider, "The provider for cloud services.  Empty string for no provider.")
 	fs.StringVar(&s.CloudConfigFile, "cloud-config", s.CloudConfigFile, "The path to the cloud provider configuration file.  Empty string for no configuration file.")
+	fs.StringVar(&s.ExternalPlugin, "external-plugin", s.ExternalPlugin, "The plugin to use when cloud provider external. Can be empty, should only be set when cloud-provider is external.")
```
Comment (wlan0) [inline cmd/kube-controller-manager/app/options/options.go]: Volume plugin to use when Cloud provider is external. Eg. 
`--external-plugin=aws` (for ebs )
`--external-plugin=gce` (for pd) etc.

**Comment by wlan0** [comment]
@cheftako Thanks for the PR. Initializing external plugin using the cloud initializer will start the cloud dependent controllers. If the cloudprovider is not null, then the controllers will start. 

For eg. The service controller uses the `cloud==nil` check to decide if it should run or not - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L210

With the current PR, `cloud` variable will not be nil if we set `external-plugin`

**Review by luxas — COMMENTED** [review]
What is _actually_ different when you specify `--external-plugin`? To me it seems like the behavior of the kcm would be the same...

Can you please explain further why this is needed?

**Comment by cheftako** [comment]
Hi @luxas as @wlan0 mentioned there is a problem in this fix which I am working on fixing but may be making it tougher to understand the point of the fix. As the cloud provider initiative (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cloud-provider-refactoring.md) proceeds kcm should not be running the following controller loops, node, route, service (and volume). We have broken out the first 3 loops from the kcm to the ccm based on the --cloud-provider flag being set to external. Volume is in parentheses because we are still a ways from being ready to break that out. However the cloud provider flag also controls initializing the cloud plugin. Without a working cloud plugin the volume controller loop breaks for volumes of type aws_ebs, gce_pd etc. So the idea is to separate control of which loops are being run from which plugin is being loaded. As @wlan0 points out I missed that we used the existence of the plugin to control the execution of the loops rather than controlling it directly off of the flag itself.

**Comment by luxas** [comment]
cc @shashidharatd This will make us able to use the ccm for federation and kube-anywhere :)

File: cmd/kube-controller-manager/app/options/options.go (hunk 1 above)
Comment (thockin) [reply cmd/kube-controller-manager/app/options/options.go]: Shouldn't the name say "volume" if it is about volumes?  But it's not a volume plugin, it's a cloud plugin.  Can we just make this extra extra clear?

Comment (luxas) [reply cmd/kube-controller-manager/app/options/options.go]: I was very confused as well. What @thockin said.

**Comment by thockin** [comment]
Assign back to me once it is LGTM'ed and ready for approval.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -184,7 +184,7 @@ func StartControllers(s *options.CloudControllerManagerServer, kubeconfig *restc
 		cloud.Initialize(clientBuilder)
 	}
 
-	versionedClient := client("shared-informers")
+	versionedClient := rootClientBuilder.ClientOrDie("shared-informers")
```
Comment (wlan0) [inline cmd/cloud-controller-manager/app/controllermanager.go]: Why don't we want shared-informed client to use ServiceAccount for authentication?

File: cmd/cloud-controller-manager/app/options/options.go [hunk 3]
```diff
@@ -33,10 +33,7 @@ import (
 
 // CloudControllerManagerServer is the main context object for the controller manager.
 type CloudControllerManagerServer struct {
-	componentconfig.KubeControllerManagerConfiguration
-
-	Master     string
-	Kubeconfig string
+	cmoptions.ControllerManagerServer
```
Comment (wlan0) [inline cmd/cloud-controller-manager/app/options/options.go]: :+1:

**Comment by wlan0** [comment]
@cheftako This code looks great! 

One comment about the client for shared-informers. I noticed KCM (https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L423) also does the same thing. I'm just wondering why it is this way. 

The other concern I have is about using the flag `--external-plugin`. To me, it doesn't clearly communicate that we are initializing the cloud plugin that enables volumes native to that cloud. I wonder if we need a different name or if this will do. 

@luxas @andrewsykim @jhorwit2 @prydie @rrati WDYT think about using `--external-plugin` flag?

File: cmd/kube-controller-manager/app/options/options.go [hunk 4]
```diff
@@ -55,91 +50,32 @@ func NewCMServer() *CMServer {
...
-	fs.Var(componentconfig.IPVar{Val: &s.Address}, "address", "The IP address to serve on (set to 0.0.0.0 for all interfaces)")
-	fs.BoolVar(&s.UseServiceAccountCredentials, "use-service-account-credentials", s.UseServiceAccountCredentials, "If true, use individual service account credentials for each controller.")
 	fs.StringVar(&s.CloudProvider, "cloud-provider", s.CloudProvider, "The provider for cloud services.  Empty string for no provider.")
-	fs.StringVar(&s.CloudConfigFile, "cloud-config", s.CloudConfigFile, "The path to the cloud provider configuration file.  Empty string for no configuration file.")
-	fs.BoolVar(&s.AllowUntaggedCloud, "allow-untagged-cloud", false, "Allow the cluster to run without the cluster-id on cloud instances.  This is a legacy mode of operation and a cluster-id will be required in the future.")
-	fs.MarkDeprecated("allow-untagged-cloud", "This flag is deprecated and will be removed in a future release.  A cluster-id will be required on cloud instances")
+	fs.StringVar(&s.ExternalPlugin, "external-plugin", s.ExternalPlugin, "The plugin to use when cloud provider is set to external. Can be empty, should only be set when cloud-provider is external.")
```
Comment (wlan0) [inline cmd/kube-controller-manager/app/options/options.go]: What plugins does it load other than volumes?

I would mention that in the help msg. For eg. `The Volume plugin to use when cloud provider is set to external. `

File: cmd/controller-manager/app/options/utils.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig).")
+	fs.StringVar(&s.Kubeconfig, "kubeconfig", s.Kubeconfig, "Path to kubeconfig file with authorization and master location information.")
+	fs.StringVar(&s.ContentType, "kube-api-content-type", s.ContentType, "Content type of requests sent to apiserver.")
+	fs.Float32Var(&s.KubeAPIQPS, "kube-api-qps", s.KubeAPIQPS, "QPS to use while talking with kubernetes apiserver.")
+	fs.Int32Var(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver.")
+	fs.DurationVar(&s.ControllerStartInterval.Duration, "controller-start-interval", s.ControllerStartInterval.Duration, "Interval between starting controller managers.")
+	fs.Float32Var(&s.NodeEvictionRate, "node-eviction-rate", 0.1, "Number of nodes per second on which pods are deleted in case of node failure when a zone is healthy (see --unhealthy-zone-threshold for definition of healthy/unhealthy). Zone refers to entire cluster in non-multizone clusters.")
```
Comment (jhorwit2) [inline cmd/controller-manager/app/options/utils.go]: The CCM deals with eviction? I thought that was still in the KCM only?

**Comment by luxas** [comment]
@wlan0 you suggested an other name yesterday, what was it you said?

**Comment by cheftako** [comment]
@luxas @wlan0 Was it externalCloudVolumePlugin?

**Comment by wlan0** [comment]
@cheftako yes. `--external-cloud-volume-plugin`

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 2 above)
Comment (cheftako) [reply cmd/cloud-controller-manager/app/controllermanager.go]: I would have to regenerate with the old client to get the exact details. The client did not have the permissions on one of the objects it was watching.

File: cmd/kube-controller-manager/app/options/options.go (hunk 4 above)
Comment (cheftako) [reply cmd/kube-controller-manager/app/options/options.go]: Done

File: cmd/controller-manager/app/options/utils.go (hunk 5 above)
Comment (cheftako) [reply cmd/controller-manager/app/options/utils.go]: Good point. We are in the process of putting parts of the NodeController in the CCM. For now I've put these two flags back in the KCM only options. If it turns out we do need them to be common they can move in the PR which adds the requirement.

**Comment by wlan0** [comment]
/assign @thockin 

Ready for your review

**Comment by dims** [comment]
Let's ship it! before we have to rebase :)

/lgtm

**Comment by deads2k** [comment]
@cheftako @lavalamp @sttts I think the intent of this pull is create a generic set of controller library code (https://github.com/kubernetes/kubernetes/blob/master/cmd/controller-manager/app/options/utils.go#L30), but a lot of controller specific code has leaked into it (https://github.com/kubernetes/kubernetes/blob/master/cmd/controller-manager/app/options/utils.go#L71-L74 as a for instance).

I think we should follow a pattern like we did for the genericapiserver of requiring only generic things be move into common code.  Sometimes that forces other refactors to happen first, but it keeps the generic library purer and generally exposes more problems.

A few concrete items here need followups:
 1. https://github.com/kubernetes/kubernetes/blob/master/cmd/controller-manager/app/options/utils.go#L51 is a specific type of config that include config for kube-controllers. It shouldn't be in a generic package
 2. `IncludeCloudLoops` (https://github.com/kubernetes/kubernetes/pull/52371/files#diff-2aa9cfe6a01016c1679752b0fcfc9244R274) looks off.  Rather than drive it that way, we have a way to enable and disable controllers and a way to eventually disable them by default.  That would provide clean separation and a clear migration path forward.

**Pull request merged**